go/pkg/cas/upload.go

```diff
@@ -17,11 +36,18 @@ type UploadInput struct {
 	// Ignored if Path is not empty.
 	Content []byte
 
-	// TODO(nodir): add Predicate.
+	// If not nil and returns false for a file, then it is not uploaded.
+	//
+	// The function participates in a cache key, so try to reuse the same functions.
```

Comment:
> Does the use case of isolate have many cases of duplicate traversals of directories and use of input exclusions?

Yes. We often traverse the same directories, and there is almost always a generic regexp-based filter. 
I am no sure how to avoid this.

Reply:
Ah I see. Good to know.

I'm also concerned about the practicality of users using the same predicate function. Specially since the `client` package operates on the InputExclusions struct. So naturally it will be creating new functions when calling `cas` each time an execution comes in. Or it would maintain a cache of `InputExclusion` to predicate function.

Do you think it makes sense to defer the predicate cache until we start measuring performance? We may also get more insight once `client` starts using `cas`.  Alternatively we could put more restrictions on InputExclusions to  enforce there is at most one per input type, and then we can cache the actual regexs.

Reply:
Yes, it doesn't work on closures. I've removed predicate from this CL. We can revisit this in another CL.

For context, the performance [was already measured](https://bugs.chromium.org/p/chromium/issues/detail?id=1193375#c37).

`client` package isn't the only user of this package. In our package where we'd use `cas` package, we have a single regex per `UploadItem`, so it is easy to cache functions per regex. [Here is our code](https://chromium-review.googlesource.com/c/infra/luci/luci-go/+/2837288/1/client/cmd/isolate/lib/common.go#349).

Ramy, also could you clarify what exactly is the suggestion? If we don't have predicate in the cache key, then the code will become incorrect because one `UploadItem` may have one predicate, and another might have another. So their dir nodes shouldn't be the same. If you are suggesting to move the predicate from `UploadItem` to a parameter of `Upload()`, then it means separate uploads cannot reuse each other's cache.

> Do we have use cases besides file name regexes + file mode checks?
Chrome doesn't have use cases besides file name regexes.

Reply:
Given that predicate support is removed from this CL, I assume this discussion is resolved. Unresolve otherwise.

Reply:
Just responding to the last message:

> For context, the performance was already measured.
I meant overhead of applying the regex repeatedly on duplicate directories in practice. I couldn't find a specific test for that in the bug but I might've missed it.

My suggestions were to (1) make the cache contain an unfiltered list of items (files, subdirs, symlinks) and apply the predicate before uploading not before storing in cache. So the cache becomes effectively a ReadDir cache. Option (2) is to cache the regexes assuming we enforce there is a constant number of regexes in the InputExclusions slice. I agree we can't move the predicate to `Upload()`. But generally agree this makes sense in a separate PR.